src/Compilers/CSharp/Test/Syntax/Parsing/RoleParsingTests.cs

```diff
@@ -0,0 +1,1206 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using Microsoft.CodeAnalysis.CSharp.Test.Utilities;
+using Microsoft.CodeAnalysis.Test.Utilities;
+using Roslyn.Test.Utilities;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace Microsoft.CodeAnalysis.CSharp.UnitTests;
+
+public sealed class RoleParsingTests : ParsingTests
+{
+    private new SyntaxTree UsingTree(string text, params DiagnosticDescription[] expectedErrors)
+        => UsingTree(text, TestOptions.RegularNext, expectedErrors);
+
+    private void UsingNode(string text, CSharpParseOptions? options = null, DiagnosticDescription[]? expectedParsingDiagnostics = null, DiagnosticDescription[]? expectedBindingDiagnostics = null)
+    {
+        options ??= TestOptions.RegularPreview;
+        expectedParsingDiagnostics ??= Array.Empty<DiagnosticDescription>();
+        expectedBindingDiagnostics ??= expectedParsingDiagnostics;
+
+        var tree = UsingTree(text, options, expectedParsingDiagnostics);
+        Validate(text, (CSharpSyntaxNode)tree.GetRoot(), expectedParsingDiagnostics);
+
+        var comp = CreateCompilation(tree);
+        comp.VerifyDiagnostics(expectedBindingDiagnostics);
+    }
+
+    public RoleParsingTests(ITestOutputHelper output) : base(output) { }
+
+    [Theory, CombinatorialData]
+    public void RoleParsing(bool isExtension)
+    {
+        Assert.True(SyntaxFacts.IsTypeDeclaration(isExtension ? SyntaxKind.ExtensionDeclaration : SyntaxKind.RoleDeclaration));
+
+        var keyword = isExtension ? "extension" : "role";
```

Comment:
>isExtension ? "extension" : "role"; [](http://example.com/codeflow?start=22&length=35)

Minor: If we used the same length tokens for each case, could we use a single set of errors, in this test and others?
`var keyword = isExtension ? "extension" : "role     ";`